mscore/plugin/api/playevent.h

```diff
@@ -0,0 +1,151 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2019 Werner Schweer and others
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2
+//  as published by the Free Software Foundation and appearing in
+//  the file LICENCE.GPL
+//=============================================================================
+
+#ifndef __PLUGIN_API_PLAYEVENT_H__
+#define __PLUGIN_API_PLAYEVENT_H__
+
+#include "libmscore/noteevent.h"
+#include "elements.h"
+
+namespace Ms {
+namespace PluginAPI {
+
+class Element;
+class Note;
+
+extern Ms::Element* unwrap(Element* e);
+
+//---------------------------------------------------------
+//   PlayEvent
+//    Wrapper class for internal Ms::NoteEvent
+//
+//   This is based on the wrapper in excerpt.h.
+//---------------------------------------------------------
+
+class PlayEvent : public QObject {
+      Q_OBJECT
+      /// The relative pitch of the event.
+      /// This is added to the parent note's actual pitch.
+      /// \since MuseScore 3.3
+      Q_PROPERTY(int pitch READ pitch WRITE setPitch)
+      /// Time to turn on the note event.
+      /// \since MuseScore 3.3
+      Q_PROPERTY(int ontime READ ontime WRITE setOntime)
+      /// The length of time for the event.
+      /// \since MuseScore 3.3
+      Q_PROPERTY(int len READ len WRITE setLen)
+      /// Time note will turn off.
+      /// This value derived from onTime and len.
+      /// \since MuseScore 3.3
+      Q_PROPERTY(int offtime READ offtime)
+      /// \cond MS_INTERNAL
+
+   protected:
+      Ms::NoteEvent* ne;
+      QObject* parentNote;
```

Comment:
I spend quite some time playing chicken and egg with the gcc and msoft compilers to please them. ie., playevent.h required elements.h and vice versa. Matt's adjustments fixed these complaints.

There is more info on this in earlier comments on this PR.

-Dale

Reply:
I did not change the type of `parentNote`. I did change the type of `papinote` to `Element*`, however, because `unwrap()` takes an `Element*`, and the compiler does not know at this point that `Note` is a subclass of `Element`.

Reply:
I'm going to see if I can un-template the implementation and see if it makes things cleaner.

Still a WIP.

-Dale

Reply:
I have a working untemplated version I just now pushed. I didn't change any of the object dereferencing stuff though.

-Dale

Reply:
I also changed the PR message to remove the disclaimer about not supporting add/remove.

-Dale

Reply:
Well, now we can get rid of `unwrap()` and make `papinote` a `Note*` and call `papinote->note()`. I just tried it and it works.

Reply:
I'll take any clean up changes you can offer as a commit I can cherry-pick. 

That's how I've been picking up your stuff.

Thanks!

Reply:
I've pushed some more clean ups. Sorry if it broke your branch. I'll stop it now. :(
-Dale

Reply:
Here: https://github.com/mattmcclinch/MuseScore/commit/01610de97683bcef80cca7c479767f9eceb19859.

This essentially puts things back the way you had it before I started helping you. But because you removed the template stuff, it works now.

Reply:
I have it integrated. Thanks for your help!

Quite a journey into the bowels of the code for me.

I hope this passes muster.

-Dale

Reply:
I think you may have missed Dmitri's original point at the beginning of this conversation, which is that `parentNote` could (and maybe even should) be declared as a `Note*` instead of a `QObject*`. This would allow you to clean up the code even more, as demonstrated here: https://github.com/mattmcclinch/MuseScore/commit/0a41a580dedcb4dffd6998755b2fc6d3a3c9edc9.

This commit also adds back the forward declaration of `class Note` in the `Ms::PluginAPI` namespace in `playevent.h`, which I thought was no longer necessary, but it turns out that it is.

Reply:
Patch absorbed.
-Dale